external-crates/move/crates/move-binary-format/src/normalized.rs

```diff
@@ -213,89 +281,380 @@ pub enum Bytecode {
-pub struct Module {
-    pub file_format_version: u32,
-    pub address: AccountAddress,
-    pub name: Identifier,
-    pub dependencies: Vec<ModuleId>,
-    pub friends: Vec<ModuleId>,
-    pub structs: BTreeMap<Identifier, Struct>,
-    pub enums: BTreeMap<Identifier, Enum>,
-    pub functions: BTreeMap<Identifier, Function>,
-    pub constants: Vec<Constant>,
-}
-
-impl Module {
-    /// Extract a normalized module from a `CompiledModule`. The module `m` should be verified.
-    /// Nothing will break here if that is not the case, but there is little point in computing a
-    /// normalized representation of a module that won't verify (since it can't be published).
-    pub fn new(m: &CompiledModule) -> Self {
-        let friends = m.immediate_friends();
-        let structs = m.struct_defs().iter().map(|d| Struct::new(m, d)).collect();
-        let enums = m.enum_defs().iter().map(|d| Enum::new(m, d)).collect();
-        let dependencies = m.immediate_dependencies();
-        let constants = m
-            .constant_pool()
+impl<S> Tables<S> {
+    fn new<Pool: StringPool<String = S>>(
+        pool: &mut Pool,
+        m: &CompiledModule,
+        include_code: bool,
+    ) -> Self {
+        let mut tables = Tables {
+            empty_signature: Rc::new(vec![]),
+            signatures: Vec::new(),
+            constants: Vec::new(),
+            struct_defs: Vec::new(),
+            function_defs: Vec::new(),
+            enum_defs: Vec::new(),
+        };
+        tables.signatures = m
+            .signatures
+            .iter()
+            .map(|s| Type::signature(pool, m, s))
+            .collect();
+        tables.constants = m
+            .constant_pool
             .iter()
-            .map(|constant| Constant::new(m, constant))
+            .map(|c| Rc::new(Constant::new(pool, m, c)))
             .collect();
-        let functions = m
-            .function_defs()
+        tables.struct_defs = m
+            .struct_defs
             .iter()
-            .map(|func_def| Function::new(m, func_def))
+            .map(|s| Rc::new(Struct::new(pool, m, s)))
+            .collect();
+        tables.enum_defs = m
+            .enum_defs
+            .iter()
+            .map(|e| Rc::new(Enum::new(pool, m, e)))
+            .collect();
+        tables.function_defs = m
+            .function_defs
+            .iter()
+            .map(|f| Rc::new(Function::new(&tables, pool, m, f, include_code)))
+            .collect();
+        tables
+    }
+}
+
+impl<S: Ord> Module<S> {
+    /// Extract a normalized module from a `CompiledModule`. The module `m` should be verified,
+    /// particularly with regards to correct offsets and bounds.
+    /// If `include_code` is `false`, the bodies of the functions are not included but the
+    /// signatures will still be present.
+    pub fn new<Pool: StringPool<String = S>>(
+        pool: &mut Pool,
+        m: &CompiledModule,
+        include_code: bool,
```

Comment:
Lots of the use cases for normalized modules were just to look at signatures (function signatures and type declarations). So this is really just an efficiency flag